src/knx/tpuart_data_link_layer.h

```diff
@@ -8,19 +8,26 @@
 
 #define MAX_KNX_TELEGRAM_SIZE 263
 
+class ITpUartCallBacks
```

Comment:
I would have use a function pointer for this. But I don't have a strong opinion about it. Not sure whats better. A function pointer may be a bit faster. Any opinion on this? I would like to have callbacks handled in a uniform way across the project.

Reply:
Normally I would have used std::function for this. I have already found some implementation which do not use dynamic memory for this.
Example: https://github.com/arobenko/embxx/blob/master/embxx/util/StaticFunction.h

Reply:
This was my try: https://github.com/nanosonde/knx/blob/feat_coupler/src/knx/simple_functional.h
But I have decided not to use it as it seems to be a mess.

Reply:
So simple_functional is not used? Then it shouldn't be in the pull request (and I don't have to try to understand it ;)

Reply:
Yes, sorry. I forgot to remove it.

Reply:
Maybe we should checkout this implementation: https://github.com/arobenko/embxx/blob/master/embxx/util/StaticFunction.h
It would have to be adapted though to replace the GASSERT().

Reply:
It will still require the header #include <functional>

Reply:
Please also read this: https://arobenko.gitbooks.io/bare_metal_cpp/content/basic_needs/callback.html

Reply:
I'd rather keep it as it is. I don't really see why we should use a complicated template. The current solution is much easier to understand.

Reply:
Ok. However, after thinking about it, I would say that the implementation should be changed.
It is the responsibilty of the network layer to give an answer if a packet should be ACK'ed or not.
For a router and also for normal device. So instead of placing it in the BAU, I think there should be an interface via the network entity to ask the network layer "shallPackedBeAcked". Then there is no need for registering a callback. It is then part of the normal network layer (entity) interface towards the datalink layer.
What do you think?

Reply:
Note: If a datalink layer does not ACK packets then it simply does not ask the network layer. E.g.: IP and RF (more specific RF ready, RF multi can ACK packets)

Reply:
Feel free to change it.

Reply:
I will change this. Stay tuned.

Reply:
After thinking about it again, I will leave it as is for now as it only affects TP data link layer which uses ACKs.